main.tf

```diff
@@ -61,6 +79,29 @@ resource "azurerm_storage_account" "example" {
   location                 = azurerm_resource_group.example.location
   account_tier             = "Standard"
   account_replication_type = "GRS"
+
+  queue_properties {
+    hour_metrics {
+      version               = "1.0"
+      include_apis          = true
+      retention_policy_days = 365
+      enabled               = true
+    }
+    minute_metrics {
+      enabled               = true
+      version               = "1.0"
+      include_apis          = true
+      retention_policy_days = 365
+    }
+    logging {
+      delete                = true
```

Comment:
**Tenable.cs remediation**
 
**[MEDIUM]: Ensure SQL server's Transparent Data Encryption (TDE) protector is encrypted with Customer-managed key**
 
**Vulnerability**
Description:

Transparent Data Encryption (TDE) with Customer-managed key support provides increased transparency and control over the TDE Protector, increased security with an HSM-backed external service, and promotion of separation of duties.

With TDE, data is encrypted at rest with a symmetric key (called the database encryption key) stored in the database or data warehouse distribution. To protect this data encryption key (DEK) in the past, only a certificate that the Azure SQL Service managed could be used. Now, with Customer-managed key support for TDE, the DEK can be protected with an asymmetric key that is stored in the Azure Key Vault. The Azure Key Vault is a highly available and scalable cloud-based key store which offers central key management, leverages FIPS 140-2 Level 2 validated hardware security modules (HSMs), and allows separation of management of keys and data for additional security.

Based on business needs or criticality of data/databases hosted on a SQL server, it is recommended that the TDE protector is encrypted by a key that is managed by the data owner (Customer-managed key).

Rationale:

Customer-managed key support for Transparent Data Encryption (TDE) allows user control of TDE encryption keys and restricts who can access them and when. Azure Key Vault, Azure's cloud-based external key management system, is the first key management service where TDE has integrated support for Customer-managed keys. With Customer-managed key support, the database encryption key is protected by an asymmetric key stored in the Key Vault. The asymmetric key is set at the server level and inherited by all databases under that server.

Once TDE protector is encrypted with a Customer-managed key, it transfers entire responsibility of respective key management on to you, and hence you should be more careful about doing any operations on the particular key in order to keep data from corresponding SQL server and Databases hosted accessible.

When deploying Customer Managed Keys, it is prudent to ensure that you also deploy an automated toolset for managing these keys (this should include discovery and key rotation), and Keys should be stored in an HSM or hardware backed keystore, such as Azure Key Vault.

As far as toolsets go, check with your cryptographic key provider, as they may well provide one as an add-on to their service.
 
**Remediation**
**From Azure Console:**

1. Go to 'SQL servers'

For the desired server instance

2. Click On 'Transparent data encryption'
3. Set 'Transparent data encryption' to 'Customer-managed key'
4. Browse through your 'key vaults' to Select an existing key or create a new key in 
the Azure Key Vault.
5. Check 'Make selected key the default TDE protector'

**From Azure CLI**

Use the below command to encrypt SQL server's TDE protector with a Customer-managed key

az sql server tde-key set --resource-group <resourceName> --server <dbServerName> --server-key-type {AzureKeyVault} --kid <keyIdentifier>

source: tenablecs-